### PR TITLE
Import application before running the event loop

### PR DESCRIPTION
**Pull request by Kludex**
- Replaces #1110
- Solves #346

### CODE REVIEW

**Comment by graingert** [comment]
how does this interact with reload?

**Comment by Kludex** [comment]
@graingert With the simple application:
```python
import anyio


async def example():
    await anyio.sleep(0)
    print("All good!")


anyio.run(example)


async def app(scope, receive, send):
    pass
```
Running `uvicorn` with `reload`:

```bash
(uvicorn3.8.12) ➜  uvicorn git:(feat/import-app-before-event-loop) ✗ uvicorn main:app --reload --port 8005
INFO:     Will watch for changes in these directories: ['/Users/marcelotryle/Development/encode/uvicorn']
INFO:     Uvicorn running on http://127.0.0.1:8005 (Press CTRL+C to quit)
INFO:     Started reloader process [6724] using watchgod
All good!
INFO:     Started server process [6738]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
WARNING:  WatchGodReload detected file change in '['/Users/marcelotryle/Development/encode/uvicorn/main.py']'. Reloading...
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [6738]
All good!
INFO:     Started server process [6774]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

**Comment by Kludex** [comment]
Question for my future self... Does this work if `MongoClient` is instantiated? The event loop will differ, right? If that's the case, do we really want this change?

**Review by euri10 — APPROVED** [review]
lgtm thanks @Kludex

File: uvicorn/server.py [hunk 1]
```diff
@@ -57,6 +57,8 @@ def __init__(self, config: Config) -> None:
 
     def run(self, sockets: Optional[List[socket.socket]] = None) -> None:
         self.config.setup_event_loop()
```
Comment (graingert) [inline uvicorn/server.py:59]: I think you should call setup_event_loop as close to asyncio.run as possible

**Review by graingert — CHANGES_REQUESTED** [review]

File: uvicorn/server.py (hunk 1 above)
Comment (Kludex) [reply uvicorn/server.py:59]: In other words... 🤔

**Comment by Kludex** [comment]
I'm not able to reproduce the issue that motivated this PR.

I'm going to close this. If someone wants to try to reproduce, and come back with the proof again, we take a decision then.

**Pull request closed without merging**